---
title: "Working with biOmics package"
author: " Tiago Chedraoui Silva, Tiago Mendes"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
    number_sections: no
    toc_depth: 2
    highlight: haddock

vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dpi = 150)
```

```{r, echo = FALSE,hide=TRUE, message=FALSE}
devtools::load_all(".")
```

# Introduction

Technological advances and public databases like The Cancer Genome Atlas (TCGA), 
The Encyclopedia of DNA Elements (ENCODE), and The NIH Roadmap Epigenomics 
Mapping Consortium (Roadmap) have provided unprecedented opportunities to 
interrogate the epigenome of cultured cancer cell lines as well as normal 
and tumor tissues with high resolution. 

Markedly however, biological information 
is stored in different formats and there is no current tool to integrate the data,
highlighting an urgent need to develop bioinformatic tools and/or computational 
softwares to overcome this challenge. 

 Our expectation is that biOMICs can effectively automate search, retrieve,
 and analyze the vast (epi)genomic data currently available from TCGA, ENCODE, 
 and Roadmap, and integrate genomics and epigenomics features with researchers 
 own high-throughput data.



# Searching using ontology-based algorithm

In order to correlate their data, we created an ontology-based search algorithm 
that uses the repository of biomedical ontologies called [BioPortal](http://www.bioportal.bioontology.org) 
to correlate the metadata to an ontology nomenclature. 

The function `biOmicsSearch` accepts the following parameters:

* **term** Term to be searched. Example: 'brain' 'u87' 'heart' 'lung' etc.
* **experiment** Experiment type (Microarray, MiRNAMicroArray, RRBS, 
DNAsequencing, ExpressionArray, Firehose, ChipSeq, fiveC, ExonArray,
DNAMethylation,	MRESeq, RepliSeq, RNASeq, miRNASeq, Rampage	Others)
* **plot** Create plot that summarize the search results? Default: FALSE
* **dir.plot** Directory to save the summary plots. Default: "searchSummary"

```{r, eval = FALSE}
# get all sampples from the query and save into tcga folder
query <- biOmicsSearch(term="brain") 
```

### Dowloading the data

Using the result from `biOmicsSearch` you can download the data using the function
`biOmicsDownload`. The parameters are:

* **lines**	Output from biOmicsSearch 
* **enc.file.type**	 Type of file to  downloaed from encode database
* **rmap.file.type** Type of file to  downloaed from roadmap database


```{r, eval = FALSE}
# get all sampples from the query and save into tcga folder
query <- biOmicsSearch(term="brain") 
```

# Working with Encode 

### Search

 You can easily search tcga samples using the `encodeSearch` function.
 Using a summary of roadmap datatabase, the function works with the
 following parameters:

*  accession
*  biosample
*  assay
*  lab
*  target
*  description
*  organism

Some examples of search are shown below
```{r, eval = FALSE}
query <- encodeSearch(accession = "ENCSR602PHV")
```

Obs: As these terms needs to be exact we validate user input before searching.

### Download
You can easily download roadmap data using the `encode.download` function.
 
```{r, eval = FALSE}
# get all sampples from the query and save into tcga folder
encodeDownload(query,path="tcga",type="fastq") 
```

Obs: The function will structure the folders to save the data as:
_Path given by the user/Experiment folder_



# Working with Roadmap 

### Search
 You can easily search roadmap samples using the `roadmapSearch` function.
 Using a summary of roadmap datatabase, the function works with the 
 following parameters:

* **samples** Partial or complete name of the sample 
("heart","	heart,fetal day96 U","ES-WA7")
* **experiment** Experiments ("H3K4me1","smRNA-Seq")
* **NA.Accession**
* **accession**	GEO sample accession Ex: GSM409307
* **center** 
* **embargo.end.date** 


```{r, eval = FALSE}
# get all H1 cell line and IMR90 cell line from platforms H3K4me1 H3K4me3
query <- roadmapSearch(sample = c("H1 cell line","IMR90 cell line"),
                       experiment = c("H3K4me1","H3K4me3"))
```

Obs: As these terms needs to be exact we validate user input before searching.


### Roadmap download
You can easily download roadmap data using the `roadmapDownload` function.
 The function accepts the following parameters:
 
 * **lines** Output from the search function
 * **type** Type of data to download (bed,wig, etc) Default: all
 * **path** Folder to save the file
 
```{r, eval = FALSE}
# get all bed sampples from the query and save into rmap folder
roadmapDownload(query, type = "bed", path = "rmap") 
```

Obs: The function will structure the folders to save the data as:
_Path given by the user/Experiment name/Sample name_


### Examples


```{r, eval = FALSE}

samples <- c("ES-I3 cell line", 
             "ES-WA7 cell line", 
             "H1 BMP4 derived mesendoderm cultured cells", 
             "H1 cell line", 
             "H1 derived mesenchymal stem cells", 
             "H9 cell line", 
             "HUES48 cell line", 
             "HUES64 cell line", 
             "HUES6 cell line", 
             "adipose derived mesenchymal stem cells",
             "UCSF-4 cell line", 
             "bone marrow derived mesenchymal stem cells", 
             "hESC-derived CD56+ mesoderm cultured cells", 
             "iPS-11a cell line", "iPS-15b cell line", 
             "iPS-18a cell line", "iPS-18c cell line", 
             "iPS-20b cell line", 
             "iPS DF 19.11 cell line", 
             "iPS DF 6.9 cell line", 
             "penis foreskin fibroblast primary cells", 
             "breast, fibroblast primary cells")

experiment <- c("H3K27me3","H3K36me3","H3K4me1","H3K9me3","H3K9ac","H3K27ac")

query <- roadmapSearch(experiment = experiment, sample = samples)
roadmapDownload(query, path = "rmap", type = "bed")
```
# biomics UI 

TBD

### Session info
```{r, eval = TRUE}
sessionInfo()
```
