---
title: "TCGAbiolinks GUI"
author: "Tiago Chedraoui Silva, Houtan Noushmehr"
date: "`r doc_date()`"
vignette: >
  %\VignetteIndexEntry{TCGAbiolinksGUI User's Guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

output:
  BiocStyle::pdf_document:
    toc: yes
    number_sections: yes
    toc_depth: 2
    highlight: haddock
  BiocStyle::html_document:
    toc: true
    number_sections: false
    toc_depth: 2
    highlight: haddock

references:
- id: ref4
  title: Inferring regulatory element landscapes and transcription factor networks from cancer methylomes
  author: 
  - family: Yao, L., Shen, H., Laird, P. W., Farnham, P. J., & Berman, B. P. 
    given:
  journal: Genome biology
  DOI: "10.1186/s13059-015-0668-3"
  URL: "http://doi.org/10.1186/s13059-015-0668-3"
  volume: 16
  number: 1
  pages: 105
  issued:
    year: 2015 

- id: ref3
  title: Orchestrating high-throughput genomic analysis with Bioconductor
  author: 
  - family: Huber, Wolfgang and Carey, Vincent J and Gentleman, Robert and Anders, Simon and Carlson, Marc and Carvalho, Benilton S and Bravo, Hector Corrada and Davis, Sean and Gatto, Laurent and Girke, Thomas and others
    given:
  journal: Nature methods
  DOI: "10.1038/nmeth.3252"
  URL: "http://doi.org/10.1038/nmeth.3252"
  volume: 12
  number: 2
  pages: 115-121
  issued:
    year: 2015

- id: ref2
  title: "TCGAbiolinks: An R/Bioconductor package for integrative analysis of TCGA data."
  author: 
  - family: Colaprico A, Silva TC, Olsen C, Garofano L, Cava C, Garolini D, Sabedot T, Malta TM, Pagnotta SM, Castiglioni I, Ceccarelli M, Bontempi G and Noushmehr H.
    given:
  journal: Nucleic Acids Research
  URL: "http://doi.org/10.1093/nar/gkv1507"
  DOI: "10.1093/nar/gkv1507"
  issued:
    year: 2015 
    
- id: ref1
  title: "ComplexHeatmap: Making Complex Heatmaps"
  author: 
  - family: Zuguang Gu
    given:
  note: "R package version 1.9.3"
  URL: "https://github.com/jokergoo/ComplexHeatmap"
  issued:
    year: 2016

- id: ref5
  title: "Pathview: an R/Bioconductor package for pathway-based data integration and visualization"
  author: 
  - family: Luo and Weijun and Brouwer and Cory
    given:
  journal: Bioinformatics
  URL: "http://doi.org/10.1093/bioinformatics/btt285"
  DOI: "10.1093/bioinformatics/btt285"
  volume: 29
  number: 14
  pages: 1830-1831
  issued:
    year: 2013


- id: ref8
  title: Molecular Profiling Reveals Biologically Discrete Subsets and Pathways of Progression in Diffuse Glioma
  author: 
  - family: Ceccarelli, Michele and Barthel, Floris P and Malta, Tathiane M and Sabedot, Thais S and Salama, Sofie R and Murray, Bradley A and Morozova, Olena and Newton, Yulia and Radenbaugh, Amie and Pagnotta, Stefano M and others
    given:
  journal: Cell
  URL: "http://doi.org/10.1016/j.cell.2015.12.028"
  DOI: "10.1016/j.cell.2015.12.028"
  volume: 164
  number: 3
  pages: 550-563
  issued:
    year: 2016


- id: ref9
  title: Comprehensive molecular profiling of lung adenocarcinoma
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Nature
  URL: "http://doi.org/10.1038/nature13385"
  DOI: "10.1038/nature13385"
  volume: 511
  number: 7511
  pages: 543-550
  issued:
    year: 2014


- id: ref10
  title: Comprehensive molecular characterization of gastric adenocarcinoma
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Nature
  URL: "http://doi.org/10.1038/nature13480"
  DOI: "10.1038/nature13480"
  issued:
    year: 2014

- id: ref11
  title: Comprehensive molecular portraits of human breast tumours
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Nature
  URL: "http://doi.org/10.1038/nature11412"
  DOI: "10.1038/nature11412"
  volume: 490
  number: 7418
  pages: 61-70
  issued:
    year: 2012
  
- id: ref12
  title: Comprehensive molecular characterization of human colon and rectal cancer
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Nature
  URL: "http://doi.org/10.1038/nature11252"
  DOI: "10.1038/nature11252"
  volume: 487
  number: 7407
  pages: 330-337
  issued:
    year: 2012    

- id: ref13
  title: Genomic classification of cutaneous melanoma
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Cell
  URL: "http://doi.org/10.1016/j.cell.2015.05.044"
  DOI: "10.1016/j.cell.2015.05.044"
  volume: 161
  number: 7
  pages: 1681-1696
  issued:
    year: 2015    

- id: ref14
  title: Comprehensive genomic characterization of head and neck squamous cell carcinomas
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Nature
  URL: "http://doi.org/10.1038/nature14129"
  DOI: "10.1038/nature14129"
  volume: 517
  number: 7536
  pages: 576-582
  issued:
    year: 2015    

- id: ref15
  title: The somatic genomic landscape of chromophobe renal cell carcinoma
  author: 
  - family: Davis, Caleb F and Ricketts, Christopher J and Wang, Min and Yang, Lixing and Cherniack, Andrew D and Shen, Hui and Buhay, Christian and Kang, Hyojin and Kim, Sang Cheol and Fahey, Catherine C and others
    given:
  journal: Cancer Cell
  URL: "http://doi.org/10.1016/j.ccr.2014.07.014"
  DOI: "10.1016/j.ccr.2014.07.014"
  volume: 26
  number: 3
  pages: 319-330
  issued:
    year: 2014    


- id: ref16
  title: Comprehensive genomic characterization of squamous cell lung cancers
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Nature
  URL: "http://doi.org/10.1038/nature11404"
  DOI: "10.1038/nature11404"
  volume: 489
  number: 7417
  pages: 519-525
  issued:
    year: 2012   

- id: ref17
  title: Integrated genomic characterization of endometrial carcinoma
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Nature
  URL: "http://doi.org/10.1038/nature12113"
  DOI: "10.1038/nature12113"
  volume: 497
  number: 7447
  pages: 67-73
  issued:
    year: 2013   

- id: ref18
  title: Integrated genomic characterization of papillary thyroid carcinoma
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Cell
  URL: "http://doi.org/10.1016/j.cell.2014.09.050"
  DOI: "10.1016/j.cell.2014.09.050"
  volume: 159
  number: 3
  pages: 676-690
  issued:
    year: 2014   

- id: ref19
  title: The molecular taxonomy of primary prostate cancer
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Cell
  URL: "http://doi.org/10.1016/j.cell.2015.10.025"
  DOI: "10.1016/j.cell.2015.10.025"
  volume: 163
  number: 4
  pages: 1011-1025
  issued:
    year: 2015   
    

- id: ref20
  title: Comprehensive Molecular Characterization of Papillary Renal-Cell Carcinoma
  author: 
  - family: Linehan, W Marston and Spellman, Paul T and Ricketts, Christopher J and Creighton, Chad J and Fei, Suzanne S and Davis, Caleb and Wheeler, David A and Murray, Bradley A and Schmidt, Laura and Vocke, Cathy D and others
    given:
  journal: NEW ENGLAND JOURNAL OF MEDICINE
  URL: "http://doi.org/10.1056/NEJMoa1505917"
  DOI: "10.1056/NEJMoa1505917"
  volume: 374
  number: 2
  pages: 135-145
  issued:
    year: 2016    
    
- id: ref21
  title: Comprehensive molecular characterization of clear cell renal cell carcinoma
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Nature
  URL: "http://doi.org/10.1038/nature12222"
  DOI: "10.1038/nature12222"
  volume: 499
  number: 7456
  pages: 43-49
  issued:
    year: 2013        
    
- id: ref22
  title: Comprehensive Pan-Genomic Characterization of Adrenocortical Carcinoma
  author: 
  - family: Cancer Genome Atlas Research Network and others
    given:
  journal: Cancer Cell
  URL: "http://dx.doi.org/10.1016/j.ccell.2016.04.002"
  DOI: "10.1016/j.ccell.2016.04.002"
  volume: 29
  pages: 43-49
  issued:
    year: 2016 

- id: ref23
  title: Complex heatmaps reveal patterns and correlations in multidimensional genomic data
  author: 
  - family: Gu, Zuguang and Eils, Roland and Schlesner, Matthias
  given:
  journal: Bioinformatics
  URL: "http://dx.doi.org/10.1016/j.ccell.2016.04.002"
  DOI: "10.1016/j.ccell.2016.04.002"
  pages: "btw313"
  issued:
   year: 2016 

- id: ref24
  title: "TCGA Workflow: Analyze cancer genomics and epigenomics data using Bioconductor packages"
  author: 
  - family:  Silva, TC and Colaprico, A and Olsen, C and D'Angelo, F and Bontempi, G and Ceccarelli, M and Noushmehr, H
  given:
  journal: F1000Research
  URL: "http://dx.doi.org/10.12688/f1000research.8923.1"
  DOI: "10.12688/f1000research.8923.1"
  volume: 5
  number: 1542
  issued:
   year: 2016 

- id: ref25
  title: "TCGAbiolinks: an R/Bioconductor package for integrative analysis of TCGA data"
  author: 
  - family:  Colaprico, Antonio and Silva, Tiago C. and Olsen, Catharina and Garofano, Luciano and Cava, Claudia and Garolini, Davide and Sabedot, Thais S. and Malta, Tathiane M. and Pagnotta, Stefano M. and Castiglioni, Isabella and Ceccarelli, Michele and Bontempi, Gianluca and Noushmehr, Houtan
  given:
  journal: Nucleic Acids Research
  URL: "http://dx.doi.org/10.1093/nar/gkv1507"
  DOI: "10.1093/nar/gkv1507"
  volume: 44
  number: 8
  pages: e71
  issued:
   year: 2016 
    
    
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dpi = 600)
```

```{r, echo = FALSE,hide=TRUE, message=FALSE,warning=FALSE}
devtools::load_all(".")
```

\newpage
# Introduction

TCGAbiolinksGUI was created to help users without knowledge of programming to search, download and analyze 
TCGA data. This package offers an graphical user interface to the R/biocondcutor packages [TCGAbiolinks]( 	http://bioconductor.org/packages/TCGAbiolinks/) [@ref2] and [ELMER](http://bioconductor.org/packages/ELMER/) packages [@ref4].
Also, some other useful packages from bioconductor, such as [ComplexHeatmap](http://bioconductor.org/packages/ComplexHeatmap/) package [@ref1] has been used for data visualization.

In order to present the package we divided this vignette based in the GUI menus that
were created based on different group of analysis. The menus and sub-menus are:

* TCGA Data
    + Get TCGA data
        + Molecular data
        + Mutation data
        + Clinical data
        + Subtype data
    + Manage Summarized Experiment
* Analysis        
    * Clinical analysis
        + Survival plot
    * Epigenetic analysis
        + Differential methylation analysis
        + Volcano plot
        + Heatmap plot
        + Mean DNA methylation plot
    * Transcriptomic analysis
        + Differential expression analysis
        + Volcano plot
        + Heatmap plot
        + Enrichement analysis
    * Genomic analysis
        + Oncoprint Plot
* Integrative analysis
    + Starburst plot
    + ELMER
* Configuration
* Help documents
    + Tutorial/vignettes
    + References

After the presentation of each menu, we dedicated a section to show how to execute some of the cases of studies showed in [TCGAbiolinks paper](http://doi.org/10.1093/nar/gkv1507) and [vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/TCGAbiolinks/inst/doc/tcgaBiolinks.html) but using only the TCGAbiolinksGUI.

\newpage
## Installation TCGAbiolinksGUI

To install the package from biocondcutor repository, please, use the code below.

```{r, eval = FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("TCGAbiolinksGUI")
```

To install the package from a binary package, please, use the code below.

```{r, eval = FALSE}
# dependencies
devtools::install_github("BioinformaticsFMRP/TCGAbiolinks")
source("https://bioconductor.org/biocLite.R")
biocLite(c("pathview","clusterProfiler","ELMER"))
install.packages(c("shiny","readr","googleVis","shinydashboard"))
devtools::install_github("thomasp85/shinyFiles")
devtools::install_github("ebailey78/shinyBS", ref="shinyBS3")
devtools::install_github("daattali/shinyjs")
install.packages("~/TCGAbiolinksGUI_0.99.0_R_x86_64-pc-linux-gnu.tar.gz", repos = NULL, type = "source")
```

## Quick start

The following commands should be used in order to start the graphical user interface.

```{r, eval = FALSE}
library(TCGAbiolinksGUI)
TCGAbiolinksGUI()
```

## Citation

Please cite both TCGAbiolinks package and TCGAbiolinksGUI: 

* Colaprico A, Silva TC, Olsen C, Garofano L, Cava C, Garolini D, Sabedot T, Malta TM, Pagnotta SM, Castiglioni I, Ceccarelli M, Bontempi G and Noushmehr H. "TCGAbiolinks: an R/Bioconductor package for integrative analysis of TCGA data." Nucleic acids research (2015): gkv1507.

Other related publications to this package:

* "TCGA Workflow: Analyze cancer genomics and epigenomics data using Bioconductor packages". F1000Research [10.12688/f1000research.8923.1](http://dx.doi.org/doi:10.12688/f1000research.8923.1) [@ref24]

Also, if you have used ELMER analysis please cite:

*  Yao, L., Shen, H., Laird, P. W., Farnham, P. J., & Berman, B. P. "Inferring regulatory element landscapes and transcription factor networks from cancer methylomes." Genome Biol 16 (2015): 105.
* Yao, Lijing, Benjamin P. Berman, and Peggy J. Farnham. "Demystifying the secret mission of enhancers: linking distal regulatory elements to target genes." Critical reviews in biochemistry and molecular biology 50.6 (2015): 550-573.

\newpage

# GDC Data

## Menu: Get GDC Data

### Sub-menu: Molecular data

In this sub-menu the user can search, download and prepare GDC data.

![GDC search menu: Main window.](tcgasearch.png)

##### Advanced seach box

Based on the following filters, it will show a table with results:

* Database
* Project filter
* Data category filter
* Data type filter
* Workflow filter
* Sample type filter
* List of barcodes
* Clinical filtrer: gender, race, vital status, tumor stage


##### Download & Prepare box

After the search, the user can download the data and prepare it into an R object.
The data can be prepared into:

* [Summarized Experiment](http://www.nature.com/nmeth/journal/v12/n2/abs/nmeth.3252.html) 
[@ref3]
* Data Frame

For the Summarized Experiment object subytpe and clinical information  will be added.
The object will save with the same name from the file name field.

### Sub-menu: Subtype data


The Cancer Genome Atlas (TCGA) Research Network has reported integrated genome-wide 
studies of various diseases. We have added some of the subtypes defined by these
reports in our package. The BRCA [@ref11], COAD [@ref12], GBM [@ref8], HNSC [@ref14], KICH [@ref15], KIRC[@ref21], KIRP [@ref20], LGG [@ref8], LUAD [@ref9], 
LUSC[@ref16], PRAD[@ref19], READ [@ref12],  SKCM [@ref13], STAD [@ref10], THCA [@ref18], UCEC [@ref17] tumors have data added. 

The user can visualize the table with the information and save it as csv or an R object (rda) file.

\newpage
### Sub-menu: Clinical data

Download GDC clinical data

[![IMAGE ALT TEXT](http://img.youtube.com/vi/54NBug9ycwM/0.jpg)](http://www.youtube.com/watch?v=54NBug9ycwM "Downloading clinical example")


### Sub-menu: Mutation data
Download the Mutation Annotation Format (MAF)

\newpage
# Analysis
## Menu: Clinical analysis

### Sub-menu: Survival plot
From the clinical data download in the TCGA data menu, the user can use it to verify the survival of groups.

![Survival plot menu: Main window.](survival.png)

##### Data 
A cvs or R object (rda) file with the clinical information.

##### Parameters 

* Group column: Select the column that identifies the group of each sample.
* Legend text: Text of the legend.
* Title: Title text.
* x-axis limit: Limits the x-axis, if 0 no cut-off is considered.
* Add p-value: Add p-value to the plot.

##### Size control
Change the size of the plot

## Menu: Manage summarized Experiment object
[![IMAGE ALT TEXT](http://img.youtube.com/vi/54rP_yZlpQs/0.jpg)](http://www.youtube.com/watch?v=54rP_yZlpQs  "Downloading clinical example")



\newpage
## Menu: Epigenetic analysis

### Sub-menu: Differential methylation analysis

The user will be able to perform a Differential methylation Regions (DMR) analysis.
The output will be a file with the folowing pattern: DMR_results_GroupCol_group1_group2_pcut_1e-30_meancut_0.55.csv
Also, the summarized Experiment will be saved with all the results inside it, the new object will be saved with _result suffix.

Obs: Depending on the number of samples and the number of probes, this analysis can go from minutes to days.

![Differential methylation analysis menu: Main window.](dmr.png)

##### Data

Select a summarized Experiment object (rda)

##### Parameters control

* Cores: Selects the number of cores to be used in the analysis
* DNA methylation threshold: minimun difference of DNA methylation levels to be considered as hyper/hypomethylated
* P-value adj cut-off: minimun adjusted P-value to be considered as siginificant.
* Group column: columns with the groups to be used in the analysis
* Groups: Select at least two groups to the analyis. For example if it has three groups (g1, g2 and g3) the analysis will
output g1 vs g2, g1 vs g3 and g2 vs g3.


### Sub-menu: Volcano plot

In this sub-menu the user will be able to plot the results from Differentially methylated regions (DMR) 
analysis and differential expression analysis (DEA). 

![Volcano plot menu: Main window.](volcano.png)

##### Data 
Expected input a csv file with the following pattern:

* For expression: DEA_results_Group_subgruop1_subgroup2_pcut_0.01_logFC.cut_2.csv
* For DNA methylation: DMR_results_Group_subgruop1_subgroup2_pcut_1e-30_meancut_0.55.csv

##### Volcano options
This box will control the x-axis thresholds "Log FC threshold" for expression and "DNA methylation threshold" for DNA methylation and the y-axis thresholds "P-value adj cut-off".


##### Highligthing options  options

Checkbox option:

* Show names: Show the names of up/down regulated genes or hypo/hyper methylated probes 
* Boxed names: put names inside a box.
There is a possibility to highlight some specific genes/probes using the "Highlightining option" suboptions.

The option "points to highlight" has the possibilities:

* highlighted - show the names for only the highlighted genes/probes list
* sinigifcant - show the names for only the up/down regulated genes or hypo/hyper methylated probes 
* both - show both gtoups

##### Color control
Change the color of the plot

##### Size control
Change the size of the plot

##### Other 

* Save file with results: Create a file with the same pattern as the one in the input, but with the 
new thersholds.


## Sub-menu: Mean DNA methylation plot

In this sub-menu the user will be able to plot the mean DNA methylation by groups.

![Mean DNA methylation plot menu: Main window.](meanmet.png)

##### Data
Expected input a R object (rda) file with a summarized Experiment object.

##### Parameters control

* Groups column: Select the column that will split the data into groups. 
This column is selected from the  sample matrix (accessed with colData)

* Subgroups column: Select the column that will highlight the different subgroups data in the groups. 
* Plot jitters: Show jitters
* Select y limits: Set lower/upper limits for y
* Sort method: Methods to sort the groups in the plot
* x-axis label angle: Change angle of the text in the x-axis

##### Size control
Change the size of the plot

\newpage
## Menu: Transcriptomic analysis
In this sub-menu the user will be able to perform an enrichemnt analysis and see
the biological process, cellular component, molecular function and pathways in which the genes 
select have a role.

### Sub-menu: Differential expression analysis

##### Gene expression object box

Select a summarized Experiment object (rda)

##### Normalization of genes

* Normalization of genes? Normalize the genes?
* Normalization of genes method? Options: gcContent, geneLength

##### Quantile filter of genes

* Quantile filter of genes?
* DEA test method: quantile, varFilter, filter1, filter2
* Threshold selected as mean for filtering

##### DEA analysis

* Log FC threshold: Log2FoldChange  threshold
* P-value adj cut-off: significant threshold
* Group column: group column in the summarized Experiment object
* Group 1: Group 1 for comparison
* Group 2: Group 2 for comparison
* DEA test method: options glmRT, exactTest

##### Pathway graphs

* DEA result: Select csv file create by the analysis.
* Pathway ID: plot results in a pathway graphs. See bioconductor [pathview](http://bioconductor.org/packages/pathview/) [@ref5] package.

![Pathway graphs output.](pathview.png)

[![IMAGE ALT TEXT](http://img.youtube.com/vi/MtEVe7_ULlQ/0.jpg)](http://www.youtube.com/watch?v=MtEVe7_ULlQ "Pathview graphs example")


### Sub-menu: Heatmap plot

![Heatmap plot menu: Main window.](heatmap.png)

##### Data

* Select R object (rda) file: Should receive a summarized Experiment object
* Results file: Should receive the output from the DEA or DMR analysis.

DEA result file should have the following pattern: DEA_result_groupCol_group1_group2_pcut_0.05_logFC.cut_0.csv
DMR result file should have the following pattern:  DMR_results_groupCol_group1_group2_pcut_0.05_meancut_0.3.csv


##### Genes/Probes selection

* By status: Based on the results file the user can select to see hyper/hypo methylated probes
* Text: the user can write a list of genes separeted by ";" , "," or a new line

#### Annotation options

* Columns annotation: using the data in the summarized experiment the user can annotate the heatmap.
* Sort by column: The order of the columns can be sorted by one of the selected columns
* Row annotation: Add annotation to rows


#### Other options

* Scale data: option "none", "by row","column"
* Cluster rows ?
* Cluster columns ?
* Show row names ?
* Show col names ?

#### Size control
Change the size of the plot and the number of bars to plot

### Sub-menu: Enrichment analysis

Researchers, in order to better understand the underlying biological processes, often want to retrieve a functional profile of a set of genes that might have an important role. This can be done by performing an enrichment analysis.

Given a set of genes that are up-regulated under certain conditions, an enrichment analysis will identify classes of genes or proteins that are over-represented using annotations for that gene set.

![Enrichment analysis  menu: Main window.](ea.png)


#### Gene selection
Input a list of genes by:

* Selection: The user can select by hand multiple genes from a list
* Text: the user can write a list of genes separeted by ";" , "," or a new line
* File: select a file (rda, csv, txt) with a column Gene_symbol or mRNA

#### Paramenter control

* Size of the text
* x upper limit: x-axis upper limit. If 0 no limit is used.
* Number of bar histogram to show: Maximum number of bars to plot

#### Plot selection

* Plot biological process?
* Plot Cellular Component?
* Plot Molecular function?
* Plot Pathways?

#### Colors control
Change the color of the plot

#### Size control
Change the size of the plot and the number of bars to plot

\newpage
## Menu: Genomic analysis

### Sub-menu: Oncoprint

Using oncoPrint function from the [ComplexHeatmap](http://bioconductor.org/packages/ComplexHeatmap/) package,
this sub-menu offer a way to visualize multiple genomic alteration.

![Oncorpint plot menu: Main window.](oncoprint.png)

[![IMAGE ALT TEXT](http://img.youtube.com/vi/cp1AwT8Ogmg/0.jpg)](http://www.youtube.com/watch?v=cp1AwT8Ogmg "Oncoprint example")

#### Data

* Select MAF file: Select a MAF file (.rda) with mutatation annotation information
* Select annotation file: Select a R object (rda)  with the metadata information. Columns "patient" or "bcr_patient_barcode"
should exist.
* Annotation columns: Which columns of the annotation file should be showed?
* Annotation position: where the annotation should be added? Top or bottom.

#### Gene selection

* Selection: The user can select by hand multiple genes from a list
* Text: the user can write a list of genes separeted by ";" , "," or a new line
* File: select a file (rda, csv, txt) with a column Gene_symbol or mRNA

#### Parameters control

* Remove empty columns? If a samples has no mutation it will be removed from the plot 
* Show column names? Show patient barcodes
* Show barplot annotation on rows? Show right barplot?

\newpage
# Integrative analysis

## Menu: Starburst plot

![Starburst plot menu: Main window.](starburst.png)

### Data 
Expected input a csv file with the following pattern:

* DEA result: DEA_results_Group_subgruop1_subgroup2_pcut_0.01_logFC.cut_2.csv
* DMR result: DMR_results_Group_subgruop1_subgroup2_pcut_1e-30_meancut_0.55.csv

### Thresholds control

The possible thresholds controls are:

* Log FC threshold: gene expression Log2FC threshold
* Expression FDR cut-off: gene expression  FDR cut-off (y-axis)
* Mean DNA methylation difference threshold: Mean DNA methylation difference threshold
* Methylation FDR cut-off: DNA methylation FDR cut-off (x-axis)

The options  Mean DNA methylation difference threshold and Log FC threshold
are used to circle relevant genes (genes that respect these thresholds)

### Highlight control
The possible highlight controls are:

* Show genes names: show names of significant genes
* Boxed names: show names inside a box
* Circle gnees: Circle candidate biologically significant genes

### Other option

* Save result: save results in a csv file

### Video:
[![IMAGE ALT TEXT](http://img.youtube.com/vi/_ec6Sij4MBc/0.jpg)](http://www.youtube.com/watch?v=_ec6Sij4MBc "Starburst plot example")


## Menu: ELMER
This sub-menu will help the user to perform the analysis from R/bioconductor ELMER package [@ref4].

![ELMER menu: Main window.](elmer.png)

### Sub-menu ELMER: analysis
#### Data: Create mee 

ELMER works with a mee object, that has matched DNA methylation and gene expression data and
divides the samples into two groups experiment and control.

* Select DNA methylation object: select a summarized Experiment object with the DNA methylation data
* Select expression object: select a summarized Experiment object with the gene expression data
* Group column: Select the columns with the groups. This column will be selected from the Summarized Experiment object.
* Experiment group: Select the columns with the subgroup 1 (experiment)
* Control group: Select the columns with the subgroup 2 (control)

* DNA methylation: Cut-off NA samples (%): By default, for the DNA methylation data 
will remove probes with NA values in more than 20% samples and remove the 
anottation data.

#### Data: Select mee
 Select the R object (rda) file with a mee object created in the previous step.

#### Analysis

This box has all the option for elmer functions. Please, see ELMER vignette. 

### Sub-menu ELMER: Visualize results
#### DATA

Select  the R object (rda) file with ELMER results created in the analysis step.

#### Plot
Select ELMER plot. To have access to all plot it needs that the results and mee object
are selected. Please, see ELMER vignette. 

#### Results table
Select which ELMER results in a table.  It needs that the results to be selected.

# Data input summary
```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
tabl <- "  
| Menu                    | Sub-menu                          | Button             | Data input  |
|---------------------------------|-----------------------------------|------------------------------------|-----------------------------------------------------------------------------------------------------------------------|
| Clinical analysis       | Profile Plot                      | Select file        | A table with at least two categorial columns  |
| Clinical analysis       | Survival Plot                     | Select file        | A table with at lest the following columns: days_to_death, days_to_last_followup and one column with a group |
| Epigenetic analysis     | Differential methylation analysis | Select data (.rda) | A summarizedExperiment object |
| Epigenetic analysis     | Volcano Plot                      | Select results     | A csv file with the following pattern: DMR_results_GroupCol_group1_group2_pcut_1e-30_meancut_0.55.csv  (Where GroupCol, group1, group2 are the names of the columns selected in the  DMR steps. |
| Epigenetic analysis     | Heatmap plot                      | Select file        | A summarizedExperiment object  |
| Epigenetic analysis     | Heatmap plot                      | Select results     | Same as Epigenetic analysis >Volcano Plot > Select results |
| Epigenetic analysis     | Mean DNA methylation              | Select file        | A summarizedExperiment object |
| Transcriptomic Analysis | Volcano Plot                      | Select results     | A csv file with the following pattern: DEA_results_GroupCol_group1_group2_pcut_1e-30_meancut_0.55.csv (Where GroupCol, group1, group2 are the names of the columns selected in the DEA steps. |
| Transcriptomic Analysis | Heatmap plot                      | Select file        | A summarizedExperiment object  |
| Transcriptomic Analysis | OncoPrint plot                      | Select MAF file        | A MAF file (columns needed: Hugo_Symbol,Tumor_Sample_Barcode,Variant_Type) |   |
| Transcriptomic Analysis | OncoPrint plot                      | Select Annotation file        | A file with at least the following columns: bcr_patient_barcode  |    |
|  Integrative analysis   | Starbrust plot                      | DMR result        | A csv file with the following pattern: DMR_results_GroupCol_group1_group2_pcut_1e-30_meancut_0.55.csv (Where GroupCol, group1, group2 are the names of the columns selected in the DMR steps.  |
|  Integrative analysis   | Starbrust plot                      | DEA result        | A csv file with the following pattern: DEA_results_GroupCol_group1_group2_pcut_1e-30_meancut_0.55.csv (Where GroupCol, group1, group2 are the names of the columns selected in the DEA steps.  |
|  Integrative analysis   | ELMER                      | Create mee > Select DNA methylation object         | An rda file with a sumarized Exepriemnt object |   
|  Integrative analysis   | ELMER                      | Select results > Select expression object         |  An rda file with the RNAseq data frame |   
|  Integrative analysis   | ELMER                      | Select mee         | An rda file with a mee object |   
|  Integrative analysis   | ELMER                      | Select results         | An rda file with the results of the ELMER analysis |   
"
cat(tabl) 
```

\newpage
# Cases of study 

## Case 3
[![IMAGE ALT TEXT](http://img.youtube.com/vi/RcXibjFanH4/0.jpg)](http://www.youtube.com/watch?v=RcXibjFanH4 "Case study 3")

## Case 4
[![IMAGE ALT TEXT](http://img.youtube.com/vi/EhGw1aF-2_c/0.jpg)](http://www.youtube.com/watch?v=EhGw1aF-2_c "Case study 4")


\newpage
# Session info

```{r sessionInfo, results='asis', echo=FALSE}
toLatex(sessionInfo())
```
\newpage
# References
