<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Integrative analysis</title>

<script src="integrative_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="integrative_files/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="integrative_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="integrative_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="integrative_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="integrative_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="integrative_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="integrative_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="integrative_files/navigation-1.1/tabsets.js"></script>
<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 70px;
}
p code {
  white-space: inherit;
}
.panel-heading {
  text-align: center;
}
</style>
<link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">


<script src="libs/jquery/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap/css/flatly.min.css" rel="stylesheet" />
<script src="libs/bootstrap/js/bootstrap.min.js"></script>
<script src="libs/bootstrap/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap/shim/respond.min.js"></script>
<script src="libs/htmlwidgets/htmlwidgets.js"></script>
<script src="libs/datatables-binding/datatables.js"></script>
<link href="libs/dt-core/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core/js/jquery.dataTables.min.js"></script>
<link href="libs/nouislider/jquery.nouislider.min.css" rel="stylesheet" />
<script src="libs/nouislider/jquery.nouislider.min.js"></script>
<link href="libs/selectize/selectize.bootstrap3.css" rel="stylesheet" />
<script src="libs/selectize/selectize.min.js"></script>
<style type="text/css">
<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="include/site.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->
<script src="libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>




<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="include/site.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<nav class="navbar navbar-default navbar-fixed-top">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a href="#" class="navbar-left"><img  src="include/symbol.png" height="42" width="40"vspace=10 hspace=10"></a>
      <a class="navbar-brand" href="./">
          Introduction
      </a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li><a href="data.html">GDC Data</a></li>
        <li><a href="analysis.html">Analysis</a></li>
        <li><a href="integrative.html">Integrative analysis</a></li>
        <li><a href="Cases.html">Case studies</a></li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
            <a href="https://github.com/BioinformaticsFMRP/TCGAbiolinksGUI" title="View Source on Github"><i class="fa fa-github fa-lg"></i>
            </a>
        </li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</nav><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Integrative analysis</h1>

</div>


<div id="menu-starburst-plot" class="section level1">
<h1><span class="header-section-number">1</span> Menu: Starburst plot</h1>
<div class="figure">
<img src="starburst.png" alt="Starburst plot menu: Main window." />
<p class="caption">Starburst plot menu: Main window.</p>
</div>
<div id="data" class="section level2">
<h2><span class="header-section-number">1.1</span> Data</h2>
<p>Expected input is a csv file with the following pattern:</p>
<ul>
<li>DEA result: DEA_results_Group_subgruop1_subgroup2_pcut_0.01_logFC.cut_2.csv</li>
<li>DMR result: DMR_results_Group_subgruop1_subgroup2_pcut_1e-30_meancut_0.55.csv</li>
</ul>
</div>
<div id="thresholds-control" class="section level2">
<h2><span class="header-section-number">1.2</span> Thresholds control</h2>
<p>The possible thresholds controls are:</p>
<ul>
<li>Log FC threshold: gene expression Log2FC threshold</li>
<li>Expression FDR cut-off: gene expression FDR cut-off (y-axis)</li>
<li>Mean DNA methylation difference threshold: Mean DNA methylation difference threshold</li>
<li>Methylation FDR cut-off: DNA methylation FDR cut-off (x-axis)</li>
</ul>
<p>The options Mean DNA methylation difference threshold and Log FC threshold are used to circle genes which pass the cut-offs previously defined (eg. mean methylation or FDR).</p>
</div>
<div id="highlight-control" class="section level2">
<h2><span class="header-section-number">1.3</span> Highlight control</h2>
<p>The possible highlight controls are:</p>
<ul>
<li>Show gene names: show names of significant genes</li>
<li>Boxed names: show names inside a box</li>
<li>Circle genes: Circle candidate biologically significant genes</li>
</ul>
</div>
<div id="other-option" class="section level2">
<h2><span class="header-section-number">1.4</span> Other option</h2>
<ul>
<li>Save result: save results in a csv file</li>
</ul>
</div>
<div id="tutorial-video" class="section level2">
<h2><span class="header-section-number">1.5</span> Tutorial video:</h2>
<center>
<p><a href="http://www.youtube.com/watch?v=_ec6Sij4MBc" title="Tutorial video: Integrating DMR analysis and DEA results to be visualized into a starburst plot - (http://www.youtube.com/watch?v=_ec6Sij4MBc)"><img src="http://img.youtube.com/vi/_ec6Sij4MBc/0.jpg" alt="IMAGE ALT TEXT" /></a></p>
<strong>Tutorial Video:</strong> Integrating DMR analysis and DEA results to be visualized into a starburst plot - (<a href="http://www.youtube.com/watch?v=_ec6Sij4MBc" class="uri">http://www.youtube.com/watch?v=_ec6Sij4MBc</a>)
</center>
</div>
</div>
<div id="menu-elmer-enhancer-linking-by-methylationexpression-relationship" class="section level1">
<h1><span class="header-section-number">2</span> Menu: ELMER (Enhancer Linking by Methylation/Expression Relationship)</h1>
<p>This sub-menu will help the user to perform an integrative analysis between DNA methylation and Gene expression using the R/Bioconductor ELMER package <span class="citation">(L Yao et al. 2015)</span>.</p>
<div id="introduction" class="section level2">
<h2><span class="header-section-number">2.1</span> Introduction</h2>
<p>Recently, many studies suggest that enhancers play a major role as regulators of cell-specific phenotypes leading to alteration in transcriptomes realated to diseases <span class="citation">(Giorgio et al. 2015; Gröschel et al. 2014; Sur et al. 2012; Lijing Yao, Berman, and Farnham 2015)</span>. In order to investigate regulatory enhancers that can be located at long distances upstream or downstream of target genes Bioconductor offer the <a href="http://bioconductor.org/packages/ELMER/">Enhancer Linking by Methylation/Expression Relationship (ELMER)</a> package. This package is designed to combine DNA methylation and gene expression data from human tissues to infer multi-level cis-regulatory networks. It uses DNA methylation to identify enhancers and correlates their state with expression of nearby genes to identify one or more transcriptional targets. Transcription factor (TF) binding site analysis of enhancers is coupled with expression analysis of all TFs to infer upstream regulators. This package can be easily applied to TCGA public available cancer data sets and custom DNA methylation and gene expression data sets <span class="citation">(L Yao et al. 2015)</span>.</p>
<p><a href="http://bioconductor.org/packages/ELMER/">ELMER</a> analysis have 5 main steps:</p>
<ol style="list-style-type: decimal">
<li><p>Identify distal enhancer probes on HM450K.</p></li>
<li><p>Identify distal enhancer probes with significantly different DNA methyaltion level in control group and experiment group.</p></li>
<li><p>Identify putative target genes for differentially methylated distal enhancer probes.</p></li>
<li><p>Identify enriched motifs for the distal enhancer probes which are significantly differentially methylated and linked to putative target gene.</p></li>
<li><p>Identify regulatory TFs whose expression associate with DNA methylation at motifs.</p></li>
</ol>
</div>
<div id="sub-menu-elmer-analysis" class="section level2">
<h2><span class="header-section-number">2.2</span> Sub-menu ELMER: analysis</h2>
</div>
<div id="data-create-mee" class="section level2">
<h2><span class="header-section-number">2.3</span> Data: Create mee</h2>
<p>The <a href="http://bioconductor.org/packages/ELMER/">ELMER</a> input is a mee object that contains a DNA methylation matrix, an gene expression matrix, a probe information GRanges, the gene information GRanges and a data frame summarizing the data. It should be highlighted that samples without both the gene expression and DNA methylation data will be removed from the mee object.</p>
<p>By default the function <em>fetch.mee</em> that is used to create the mee object will separate the samples into two groups, the control group (normal samples) and the experiment group (tumor samples), but the user can relabel the samples to compare different groups.</p>
<p>The following steps are required:</p>
<ul>
<li>Select DNA methylation object: select a summarized Experiment object with the DNA methylation data</li>
<li>Select expression object: select a summarized Experiment object with the gene expression data</li>
<li>Select group column: Select the columns with the groups. This column will be selected from the Summarized Experiment object.</li>
<li>Select Experiment group: Select the columns with the subgroup 1 (experiment)</li>
<li>Select Control group: Select the columns with the subgroup 2 (control)</li>
<li>Select DNA methylation: Cut-off NA samples (%): By default, for the DNA methylation data will remove probes with NA values in more than 20% samples and remove the annotation data.</li>
</ul>
<p>Also, for RNA expression data it will take the log2(expression + 1) of the expression matrix in order to linearize the relation between DNA methylation and expression.</p>
<div class="figure">
<img src="elmer.png" alt="ELMER menu: Main window." />
<p class="caption">ELMER menu: Main window.</p>
</div>
<div id="data-select-mee" class="section level3">
<h3><span class="header-section-number">2.3.1</span> Data: Select mee</h3>
<p>Select the R object (rda) file with a mee object created in the previous step.</p>
</div>
<div id="analysis-parameters" class="section level3">
<h3><span class="header-section-number">2.3.2</span> Analysis parameters</h3>
<p>After preparing the data into a mee object, we will execute the five <a href="http://bioconductor.org/packages/ELMER/">ELMER</a> steps for both the hypo (distal enhancer probes hypomethylated in the experiment group) and hyper (distal enhancer probes hypermethylated in the experiment group) direction.</p>
<p>This box has all the available options for ELMER functions. Please see the <a href="http://bioconductor.org/packages/release/bioc/vignettes/ELMER/inst/doc/vignettes.pdf">ELMER vignette</a>.</p>
<p>Also, a description of the data used by ELMER (such as the distal enhacer probes) is found in the <a href="https://bioconductor.org/packages/release/data/experiment/vignettes/ELMER.data/inst/doc/vignettes.pdf">ELMER.data</a> vignette.</p>
<p>ELMER Identifies the enriched motifs for the distal enhancer probes which are significantly differentially methylated and linked to putative target gene, it will plot the Odds Ratio (x axis) for the each motifs found. These motifs by default have a minimum incidence of 10 probes (that means at least 10 probes were associated with the motif) in the given probes set and the smallest lower boundary of 95% confidence interval for Odds Ratio of 1.1.</p>
<p>After finding the enriched motifs, <a href="http://bioconductor.org/packages/ELMER/">ELMER</a> identifies regulatory transcription factors (TFs) whose expression is associated with DNA methylation at motifs. <a href="http://bioconductor.org/packages/ELMER/">ELMER</a> automatically creates a TF ranking plot for each enriched motif. This plot shows the TF ranking plots based on the association score <span class="math inline">( − <em>l</em><em>o</em><em>g</em>(<em>P</em><em>v</em><em>a</em><em>l</em><em>u</em><em>e</em>))</span> between TF expression and DNA methylation of the motif. We can see in Figure below that the top three TFs that are associated with a motif found.</p>
<p>In case, it identifies regulatory transcription factors (TFs), a object with the prefix ELMER_results will be created with the necessary data to visualize the results.</p>
</div>
</div>
<div id="sub-menu-elmer-visualize-results" class="section level2">
<h2><span class="header-section-number">2.4</span> Sub-menu ELMER: Visualize results</h2>
<div id="data-1" class="section level3">
<h3><span class="header-section-number">2.4.1</span> Data</h3>
<p>Select the R object (rda) file with ELMER results created in the analysis step (the one with prefix ELMER_results )</p>
</div>
<div id="plot" class="section level3">
<h3><span class="header-section-number">2.4.2</span> Plot</h3>
<p>Select ELMER plot. To access the plots, both the results and mee object must be selected. Please see <a href="http://bioconductor.org/packages/release/bioc/vignettes/ELMER/inst/doc/vignettes.pdf">ELMER vignette</a> for more details.</p>
</div>
<div id="results-table" class="section level3">
<h3><span class="header-section-number">2.4.3</span> Results table</h3>
<p>Select the ELMER results from the table. A results object should be selected.</p>
<div id="refs" class="references">
<div id="ref-giorgio2015large">
<p>Giorgio, Elisa, Daniel Robyr, Malte Spielmann, Enza Ferrero, Eleonora Di Gregorio, Daniele Imperiale, Giovanna Vaula, et al. 2015. “A Large Genomic Deletion Leads to Enhancer Adoption by the Lamin B1 Gene: A Second Path to Autosomal Dominant Leukodystrophy (Adld).” <em>Human Molecular Genetics</em>. Oxford Univ Press, ddv065.</p>
</div>
<div id="ref-groschel2014single">
<p>Gröschel, Stefan, Mathijs A Sanders, Remco Hoogenboezem, Elzo de Wit, Britta AM Bouwman, Claudia Erpelinck, Vincent HJ van der Velden, et al. 2014. “A Single Oncogenic Enhancer Rearrangement Causes Concomitant Evi1 and Gata2 Deregulation in Leukemia.” <em>Cell</em> 157 (2). Elsevier: 369–81.</p>
</div>
<div id="ref-sur2012mice">
<p>Sur, Inderpreet Kaur, Outi Hallikas, Anna Vähärautio, Jian Yan, Mikko Turunen, Martin Enge, Minna Taipale, Auli Karhu, Lauri A Aaltonen, and Jussi Taipale. 2012. “Mice Lacking a Myc Enhancer That Includes Human Snp Rs6983267 Are Resistant to Intestinal Tumors.” <em>Science</em> 338 (6112). American Association for the Advancement of Science: 1360–3.</p>
</div>
<div id="ref-yao2015inferring">
<p>Yao, L, H Shen, PW Laird, PJ Farnham, and BP Berman. 2015. “Inferring Regulatory Element Landscapes and Transcription Factor Networks from Cancer Methylomes.” <em>Genome Biology</em> 16 (1): 105–5.</p>
</div>
<div id="ref-yao2015demystifying">
<p>Yao, Lijing, Benjamin P Berman, and Peggy J Farnham. 2015. “Demystifying the Secret Mission of Enhancers: Linking Distal Regulatory Elements to Target Genes.” <em>Critical Reviews in Biochemistry and Molecular Biology</em> 50 (6). Taylor &amp; Francis: 550–73.</p>
</div>
</div>
</div>
</div>
</div>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
    // active menu
    var href = window.location.pathname;
    href = href.substr(href.lastIndexOf('/') + 1);
    if (href === '') href = './';
    $('a[href="' + href + '"]').parent().addClass('active');
});
</script>
<script>
// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>


</body>
</html>
